### PR TITLE
Sema: NSValue-to-value-type casts are failable and should be checked.

### PR DESCRIPTION
**Pull request by jckarter**
In Swift 4 mode, no longer consider e.g. 'nsNumber as Int' or 'nsValue as NSRange' to be valid coercions. This would break compatibility with Swift 3, so in Swift 3 mode, accept the coercion, but *also* accept a checked cast without a warning, and raise a migration warning about the unchecked coercion.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@DougGregor and @rudkx, do you mind reviewing this?

**Review by rudkx — APPROVED** [review]
Everything else LGTM.

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -5394,6 +5411,8 @@ Expr *ExprRewriter::buildObjCBridgeExpr(Expr *expr, Type toType,
                                         ConstraintLocatorBuilder locator) {
   Type fromType = cs.getType(expr);
 
+  // Warn in Swift 3 mode if a coercion goes from
```
Comment (jckarter) [inline lib/Sema/CSApply.cpp]: Fixed, thanks Mark!

**Comment by jckarter** [comment]
@swift-ci Please smoke test and merge

**Comment by jckarter** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**